### ISSUE TITLE
Logging causes Hammerspoon slowdown over time

### ISSUE THREAD

**Issue by eccentric-j** [Hammerspoon/hammerspoon#3389]
Co-maintainer of https://github.com/agzam/spacehammer

We've had reports over the past year or so about things like hs.alert opening very slowly and the hotkey modals taking a while to start and stop before switching to the next one.

A day or so ago accidentally discovered that issue was tied to logging upstream in Hammerspoon. The more that is logged over time, the slower hammerspoon becomes.

## Reproduction:

1. Create a binding that opens hs.alert and then create another binding that closes hs.alert
2. Fill up the log. Sharing my fennel test code below, but would recommend doing it asynchronously in a timer.
3. Before the logging starts, open and close the alert. Should happen nearly instantly.
4. After logging starts for a while, open and close the alert. Depending on how much info was logged, the UI will open and close a second or more after pressing the binding.

### Expected:

1. Ideally no matter how much info is logged, wouldn't expect a runtime performance decrease

### Actual

The more info that is logged, the slower Hammerspoon performs all around

Here's me opening the alert right after clearing the log:

![2023-02-22 20 15 58](https://user-images.githubusercontent.com/590297/221030075-4aff082c-9951-4d20-bebd-26bbc439d75e.gif)

Notice the time between pressing "m" and when my media modal menu opens.

![2023-02-22 20 22 55](https://user-images.githubusercontent.com/590297/221031561-29d8f2f9-c9d9-4da2-8b93-8c8bc2dbd234.gif)

- Pressing the same binding after artificially stuffing the log, the time between when "m" is pressed vs when the media modal opening has noticeably slowed down to about a second
- Clearing the hammerspoon log, then invoking the binding it's fast again.

My test code, specifics are not important here it just logs more data each iteration at a slower rate so the system is responsive. Use whatever you want to test.

```fennel
(local state
  {:loop-count 0
   :delay      10
   :entries    []})

(fn start!
  []
  (let [tick (fn [f]
               (hs.timer.doAfter
                 (/ state.delay 1000)
                (fn []
                  (print "debug-modal-delay"
                         (hs.inspect state))
                  (table.insert state.entries (.. "" (os.time)))
                  (set state.loop-count (+ state.loop-count 1))
                  (when (= (% state.loop-count 100) 0)
                    (set state.delay (* state.delay 2)))
                  (f f))))]
    (tick tick)))

{: start!}
```

Any thoughts as to the exact cause? Admittedly I'm not familiar with Obj-C\Swift but happy to help find and\or fix in any way I can.

**Comment by Grazfather** [comment]
@eccentric-j and I dug into the code base and found that the console `textStorage` is a simple `NSMutableAttributedString`, which we append to on print and console log. It makes sense, then, that this would slow down as the string gets huge, as appending in O(n), and we might suffer costly reallocs.

Maybe the console log should be converted to use some sort of circular buffer?

**Comment by Grazfather** [comment]
Maybe we just set `maximumNumberOfLines`? That might do the trick.

**Comment by cmsj** [comment]
Do we know what happens if one exceeds `maximumNumberOfLines`? Apple's API docs don't say

**Comment by eccentric-j** [comment]
I wasn't sure myself. The docs are a bit terse on that property but according to chatgpt:

> If the text exceeds the maximumNumberOfLines limit, the system will truncate the text by removing characters from the last line until it fits within the container's bounds. The removed characters will be replaced with an ellipsis character (...) to indicate that the text has been truncated.

Which seems to be confirmed by https://nickharris.wordpress.com/2016/09/11/limiting-the-number-of-lines-in-a-uitextview/

If that's true, that does not sound like that's what we want. Maybe we should find a TextContainer implementation with a circular buffer so that the first text is removed to make room for the new lines? I think that's how terminals usually work?

**Comment by latenitefilms** [comment]
Maybe something like this?

https://stackoverflow.com/questions/39605458/limit-text-length-of-auto-scrolling-nstextview

**Comment by latenitefilms** [comment]
I wonder if this change would have any impact?

Replace:

```objc
[[self.outputView textStorage] performSelectorOnMainThread:@selector(appendAttributedString:)
                                       withObject:attrstr
                                    waitUntilDone:YES];
```

...with:

```objc
    [self.outputView.textStorage beginEditing];
    [self.outputView.textStorage appendAttributedString:attrstr];
    [self.outputView.textStorage endEditing];
    [self.outputView didChangeText];
```

**Comment by cmsj** [comment]
@latenitefilms Couple of objections to that - firstly I don't remember if we are guaranteed to only ever call `appendString:type:` on the main thread, we'd have to thoroughly audit that before removing the `performSelectorOnMainThread` wrap, and secondly I don't think it would have any impact on performance - the same work is being done, we'd just be changing when it's done.

I think it's far more likely that the performance drop is because the underlying `textStorage` object becomes enormous. I've pushed up #3400 (currently untested) which would implement a maximum length on the `textStorage` and remove stuff from the beginning accordingly.

**Comment by asmagill** [comment]
FWIW, a slowdown because of a huge console backlog *is* one of the reasons `hs.console.clearConsole()` was added.

I can confirm that the few times I've spewed enough crap to the console that it affected responsiveness, clearing the console took care of it, so I suspect @cmsj's suggestion above (purging earlier data) will work.

I would suggest that we make it something that can be adjusted at runtime (say something like `hs.console.maxLines(x)` -- there have been a handful of occasions when I've intended to create a particularly large output stream during testing or exploration.

**Comment by cmsj** [comment]
@asmagill yep, the PR I pushed up, #3400 does that, although not on a line basis, because afaict the underlying NSTextStorage doesn't think about lines

**Comment by latenitefilms** [comment]
FWIW - I read somewhere that adding beginEditing/endEditing fixes a performance issue - at least on iOS. I pasted in some enormous amount of text into the Console and Hammerspoon still seemed responsive at least on my machine.

**Comment by eccentric-j** [comment]
Something that crossed my mind that I need to test is that the slowdown might not actually affect hammerspoon itself, but future logging. For example if you log data in a loop before performing an action like `hs.alert`, and the text becomes enormous, then you experience the slowdown waiting for the logging operation to finish before `hs.alert` is called. 

I'll try removing log statements from our project and setting the global log level to hide as much as possible and see how that goes.

**Comment by latenitefilms** [comment]
Also, maybe of interest:

> You can access a text storage object from any thread of your app, but your app must guarantee access from only one thread at a time.

**Comment by eccentric-j** [comment]
Thanks @cmsj and everyone we really appreciate it!
